tools/meteor-services/deploy.js

```diff
@@ -321,6 +330,120 @@ function canonicalizeSite(site) {
+// We envision potentially creating this configuration object in a programmatic
+// way or via user-specification in the future.
+// Default initialWaitTime is 10 seconds – this is the time to wait before checking at all
+// Default pollInterval is 700 milliseconds – this is the wait interval between polls
+// Default timeout is 15 minutes
+// `start` tracks the time when we started polling
+// `currentMessage` tracks what the current status message is for this version
+class PollingState {
+  constructor(timeoutMs,
+    initialWaitTimeMs,
+    pollIntervalMs,
+    maxErrors) {
+      const FIFTEEN_MINUTES_MS = 15*60*1000;
+      const MAX_ERRORS = 5
+      this.initialWaitTimeMs = initialWaitTimeMs || 10*1000;
+      this.pollIntervalMs = pollIntervalMs || 700;
+      this.deadline = timeoutMs ? new Date(new Date().getTime() + timeoutMs) :
+        new Date(new Date().getTime() + FIFTEEN_MINUTES_MS);
+      this.start = new Date();
+      this.currentMessage = '';
+      this.errors = 0;
+      this.maxErrors = maxErrors || MAX_ERRORS;
+  }
+}
+
+// Poll the "version-status" endpoint for the build and deploy status
+// of a specified version ID with a polling configuration.
+// This will only end successfully when the polling endpoint reports that
+// the version deployment is finished. The version-status endpoints will report
+// messages pertaining to the status of the version, which will then be reported
+// directly to the user. When the poll is complete, it will return an object
+// with information about the final state of the version and the app.
+function pollForDeploy(pollingState, versionId, galaxyUrl) {
+  const {
+    deadline,
+    initialWaitTimeMs,
+    pollIntervalMs,
+    start,
+    currentMessage,
+  } = pollingState;
+
+  // Do a call to the version-status endpoint for the specified versionId
+  const versionStatusResult = deployRpc({
+    method: 'GET',
+    operation: 'version-status',
+    site: versionId,
+    expectPayload: ['message', 'finishStatus'],
+    printDeployURL: false,
+  });
+
+  //Check the details of the Version Status response and compare message to last call
+  if(versionStatusResult &&
+    versionStatusResult.payload &&
+    versionStatusResult.payload.message) {
+      const message = versionStatusResult.payload.message;
+      if (currentMessage !== message) {
+        Console.info(message);
+        pollingState.currentMessage = message;
+      }
+  } else {
+    // If we did not get a valid Version Status response, just fail silently and
+    // keep polling as per usual – this may have just been a whiff from Galaxy
+    // We do the retry here because we might hit an error if we try to parse the
+    // result of the version-status call below.
+    Console.warn('Received malformed response from Galaxy',
+      versionStatusResult.errorMessage);
+    pollingState.errors++;
+    if (pollingState.errors >= pollingState.maxErrors) {
+      throw new Error(versionStatusResult.errorMessage);
```

Comment:
This will end up as a stack trace, right? In general we try to avoid printing stack traces in command-line tools for things that aren't bugs in the program itself.